### PR TITLE
fix #13115

### LINKED ISSUE #13115: newException terminates msg early from NULL byte, other IO does not
Error output produced by `raise newException` is inconsistent with other IO such as `echo` or `stdout.write`.

### Example
```nim
const b_null: char = 0.char
var msg = "This char is `" & $b_null & "` and works fine!"

echo msg

raise newException(Exception, msg)
```

### Current Output
```nim
echo msg
# This char is `` and works fine!

raise newException(Exception, msg)
# Error: unhandled exception: This char is `
```

### Expected Output
```nim
echo msg
# This char is `` and works fine!

raise newException(Exception, msg)
# Error: unhandled exception: This char is `` and works fine! [Exception]
```

### Additional Information

* This issue was found via fuzzing some CLI utility, which takes piped input. The error messsage was being cutoff, because the input contained a NULL byte.
EDIT:
* Consistency is the main thing I'd ask for, however, if both could output `^@` for NULL char, that may be better assuming it doesn't introduce needless complexity. 

```
$ nim -v
Nim Compiler Version 1.0.4 [Linux: amd64]
Compiled at 2019-11-27
Copyright (c) 2006-2019 by Andreas Rumpf

git hash: c8998c498f5e2a0874846eb31309e1d1630faca6
active boot switches: -d:release

```

### PR DESCRIPTION
**Pull request by ringabout**
It works for c backend with debug and release mode.
It doesn't work for cpp backend.

### CODE REVIEW

**Comment by ringabout** [comment]
I am not sure how `testament` handle `\0` in output strings. I will look at that first.

**Comment by ringabout** [comment]
It works for `noCppExceptions`
`koch temp cpp -r -d:noCppExceptions -d:release --nimcache:nimcache test.nim`

And `--gc:arc` also works in c backend.
`koch temp c -r -d:release --gc:arc --nimcache:nimcache test.nim`

Got:
```
Hint: D:\QQPCmgr\Desktop\Nim\test.exe  [Exec]
test.nim(4)              test
Error: unhandled exception: This char is ` ` and works fine! [Exception]
Error: execution of an external program failed: 'D:\QQPCmgr\Desktop\Nim\test.exe '
Traceback (most recent call last)
```

File: tests/exception/t13115.nim [hunk 1]
```diff
@@ -0,0 +1,13 @@
+discard """
+  exitcode: 1
+  targets: "c"
+  matrix: "-d:debug; -d:release"
+  outputsub: '''t13115.nim(13)           t13115
+Error: unhandled exception: This char is'''
+  outputsub: ''' and works fine! [Exception]'''
+"""
+
+const b_null: char = 0.char
```
Comment (timotheecour) [inline tests/exception/t13115.nim:10]: const bNull = 0.char

File: tests/exception/t13115.nim [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+var msg = "This char is `" & $b_null & "` and works fine!"
+
+raise newException(Exception, msg)
```
Comment (timotheecour) [inline tests/exception/t13115.nim:13]: or even:
```nim
raise newException(Exception, "foo" & '\0' & "bar")
```

File: tests/exception/t13115.nim [hunk 3]
```diff
@@ -0,0 +1,13 @@
+discard """
+  exitcode: 1
+  targets: "c"
+  matrix: "-d:debug; -d:release"
+  outputsub: '''t13115.nim(13)           t13115
+Error: unhandled exception: This char is'''
+  outputsub: ''' and works fine! [Exception]'''
```
Comment (timotheecour) [inline tests/exception/t13115.nim:7]: @xflywind there's a bug here, only the last outputsub matter.

testament should give error when outputsub is set twice (or better, allow it to do the right thing)
see
```nim
      of "outputsub":
        result.outputCheck = ocSubstr
        result.output = strip(e.value)
```

Comment (ringabout) [reply tests/exception/t13115.nim:7]: I can't use `output`. The behaviour differs handling `\0` in different platforms in testament. Maybe it is the bug of testament.
If only last outputsup matters, the test is still right.

File: tests/exception/t13115.nim (hunk 1 above)
Comment (timotheecour) [reply tests/exception/t13115.nim:10]: =>https://github.com/nim-lang/Nim/pull/16166

File: tests/exception/t13115.nim (hunk 2 above)
Comment (timotheecour) [reply tests/exception/t13115.nim:13]: =>https://github.com/nim-lang/Nim/pull/16166

File: tests/exception/t13115.nim (hunk 3 above)
Comment (timotheecour) [reply tests/exception/t13115.nim:7]: my point was that the 1st outputsub you set is silently overwritten so it wouldn't matter what you would've put there.

and yes, it is a bug of testament, fixing in https://github.com/nim-lang/Nim/pull/16166

Comment (ringabout) [reply tests/exception/t13115.nim:7]: Yeah I mean `hello\0test` differs in `output` with `testament`
In windows
```nim
hello
test
```
In linux like this or so
```nim
hellotest
```
Works fine if running it locally.

Comment (ringabout) [reply tests/exception/t13115.nim:7]: Like this
```
Failure: reOutputsDiffer
Expected:
t13115.nim(13)           t13115
Error: unhandled exception: This char is `
` and works fine! [Exception]

Gotten:
t13115.nim(13)           t13115
Error: unhandled exception: This char is `` and works fine! [Exception]
```

File: lib/system/ansi_c.nim [hunk 4]
```diff
@@ -154,6 +154,11 @@ proc c_fwrite*(buf: pointer, size, n: csize_t, f: CFilePtr): cint {.
 proc c_fflush(f: CFilePtr): cint {.
   importc: "fflush", header: "<stdio.h>".}
 
+proc rawWriteString*(f: CFilePtr, s: cstring, length: int) {.compilerproc, nonReloadable, inline.} =
+  # we cannot throw an exception here!
+  discard c_fwrite(s, 1, cast[csize_t](length), f)
```
Comment (timotheecour) [inline lib/system/ansi_c.nim:159]: but then we have no way to check whether these succeeded; how about return bool for success at least?

Comment (ringabout) [reply lib/system/ansi_c.nim:159]: I don't know the reason, I just copy them from original version:
```nim
proc rawWrite*(f: CFilePtr, s: cstring) {.compilerproc, nonReloadable, inline.} =
  # we cannot throw an exception here!
  discard c_fwrite(s, 1, cast[csize_t](s.len), f)
```

**Comment by timotheecour** [comment]
> I am not sure how `testament` handle `\0` in output strings. I will look at that first.

not needed in this PR but whenever testament spec is insufficient, you can shell out to call nim directly via `execCmdEx` (as done in tests/stdlib/tosproc or tests/misc/trunner.nim) and examine the output directly in your test

**Pull request merged**